src/nsls2api/services/beamline_service.py

```diff
@@ -74,9 +74,87 @@ async def detectors(name: str) -> Optional[list[Detector]]:
     if detectors is None:
         return None
 
+    Beamline.set_detectors(detectors)
     return detectors.detectors
 
 
+async def add_detector(
+    beamline_name: str,
+    detector_name: str,
+    directory_name: str
+) -> Optional[Detector]:
+    """
+    Add a new detector to a beamline.
+
+    Args:
+        beamline_name (str): The name of the beamline.
+        detector_name (str): The name of the detector.
+        directory_name (str): The directory name of the detector.
+
+    Returns:
+        Optional[Detector]: The newly created Detector object if successful, None otherwise.
+    """
+    beamline = await Beamline.find_one(Beamline.name == beamline_name.upper())
+
+    new_detector = Detector(name=detector_name, directory_name=directory_name)
+
+    if new_detector in beamline.detectors:
```

Comment:
I'm not totally confident this is a correct check. For two reasons:
1. Isn't this checking for object identity (i.e. that both references point to the same object in memory) rather than the objects' values? AFAIK we do not implement `__eq__` for this class yet
2. Even if this were comparing strictly by value, it's still probably not sufficient. It might be OK to have duplicate detector names since a beamline may have multiple detectors of the same model (really we should extend the `Detector` class to include model information and have names that include instance identifiers, but that can wait) - but it is not OK to have duplicate directory names on multiple detectors. What if a two differently named detectors have the same directory, that might look like "objects are different by values" but really these resources would still conflict.